pkg/segment/query/iqr/iqrreader.go

```diff
@@ -0,0 +1,412 @@
+	if !iqrRdr.IsMultiReader() {
+		return fmt.Errorf("iqrReader.AddReader: reader mode is not multi reader: %v", iqrRdr.readerMode)
+	}
+
+	if rdr, ok := iqrRdr.readerIdToReader[readerId]; ok {
+		if readerId != rdr.GetReaderId() {
+			return fmt.Errorf("iqrReader.AddReader: readerId=%v already exists", readerId)
+		}
+
+		return nil
+	}
+
+	iqrRdr.readerIdToReader[readerId] = reader
+
+	return nil
+}
+
+func (iqrRdr *IQRReader) AddSegEncodingToReader(segEnc utils.T_SegEncoding, readerId utils.T_SegReaderId) error {
+	if !iqrRdr.IsMultiReader() {
+		return fmt.Errorf("iqrReader.AddSegEncodingToReader: reader mode is not multi reader: %v", iqrRdr.readerMode)
+	}
+
+	if _, ok := iqrRdr.readerIdToReader[readerId]; !ok {
+		return fmt.Errorf("iqrReader.AddSegEncodingToReader: readerId=%v not found", readerId)
+	}
+
+	if ri, ok := iqrRdr.encodingToReaderId[segEnc]; ok {
+		if ri != readerId {
+			return fmt.Errorf("iqrReader.AddSegEncodingToReader: segEnc=%v already exists for readerId=%v", segEnc, ri)
+		}
+
+		return nil
+	}
+
+	iqrRdr.encodingToReaderId[segEnc] = readerId
+
+	return nil
+}
+
+func (iqrRdr *IQRReader) AddSingleIQRReader(iqr *IQR) error {
+	if !iqrRdr.IsMultiReader() {
+		return fmt.Errorf("iqrReader.AddSingleIQRReader: reader mode is not multi reader: %v", iqrRdr.readerMode)
+	}
+
+	otherRdr := iqr.reader
+	if !otherRdr.IsSingleReader() {
+		return fmt.Errorf("iqrReader.AddSingleIQRReader: other reader is not single reader: %v", otherRdr.readerMode)
+	}
+
+	if len(iqr.encodingToSegKey) == 0 {
+		return nil
+	}
+
+	otherReaderId := otherRdr.reader.GetReaderId()
+
+	err := iqrRdr.AddReader(otherReaderId, otherRdr.reader)
+	if err != nil {
+		return fmt.Errorf("iqrReader.AddSingleIQRReader: cannot add reader to iqr; err=%v", err)
+	}
+
+	for segEnc := range iqr.encodingToSegKey {
+		if err := iqrRdr.AddSegEncodingToReader(segEnc, otherReaderId); err != nil {
+			return fmt.Errorf("iqrReader.AddSingleIQRReader: cannot add seg encoding to reader; err=%v", err)
+		}
+	}
+
+	return nil
+}
+
+func (iqr *IQR) mergeIQRReaders(otherIQR *IQR) error {
+	iqrRdr := iqr.reader
+	otherRdr := otherIQR.reader
+
+	if iqrRdr.IsSingleReader() && otherRdr.IsSingleReader() {
+		if reflect.DeepEqual(iqrRdr, otherRdr) {
+			return nil
+		}
+
+		resultRdr := NewMultiIQRReader()
+
+		err := resultRdr.AddSingleIQRReader(iqr)
+		if err != nil {
+			return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add iqr to resultRdr; err=%v", err)
+		}
+
+		err = resultRdr.AddSingleIQRReader(otherIQR)
+		if err != nil {
+			return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add otherIQR to resultRdr; err=%v", err)
+		}
+
+		iqr.reader = resultRdr
+
+		return nil
+	}
+
+	if iqrRdr.IsSingleReader() {
+		iqrRdr.readerMode = ReaderModeMultiReader
+		if err := iqrRdr.AddSingleIQRReader(otherIQR); err != nil {
+			return fmt.Errorf("iqrReader.MergeIQRReaders: cannot add otherIQR to iqr; err=%v", err)
+		}
+	}
```

Comment:
Is this right? Above, you have 
```
if iqrRdr.IsSingleReader() && otherRdr.IsSingleReader() {
    ...
    return
}
```

So on 191, the `if iqrRdr.IsSingleReader()` is the same as `if iqrRdr.IsSingleReader() && !otherRdr.IsSingleReader()`. So should you be calling `AddSingleIQRReader()` on `otherIQR`, or instead call something like `AddMultiIQRReader()`?